## 1. Investigation Overview

This project analyzes the data set that includes information about individual rides made in a bike-sharing system covering the greater San Francisco Bay area. In this project the focus will be on finding diffrent aspects of user type.




In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
import datetime as dt
import networkx as nx
import matplotlib.pyplot as plt

In [2]:
# load in the dataset into a pandas dataframe, show first three rows
df_bike = pd.read_csv('C:\\Users\\raz37388\\Desktop\\Udacity_assignment\\Assignment 4\\2017-fordgobike-tripdata.csv')
df_bike.head(3)

duration_sec                start_time                  end_time  \
0         80110  2017-12-31 16:57:39.6540  2018-01-01 15:12:50.2450   
1         78800  2017-12-31 15:56:34.8420  2018-01-01 13:49:55.6170   
2         45768  2017-12-31 22:45:48.4110  2018-01-01 11:28:36.8830   

   start_station_id                                 start_station_name  \
0                74                              Laguna St at Hayes St   
1               284  Yerba Buena Center for the Arts (Howard St at ...   
2               245                             Downtown Berkeley BART   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.776435              -122.426244              43   
1               37.784872              -122.400876              96   
2               37.870348              -122.267764             245   

                                    end_station_name  end_station_latitude  \
0  San Francisco Public Library (Grove St at Hyde...             37.778768   
1                              Dolores St at 15th St             37.766210   
2                             Downtown Berkeley BART             37.870348   

   end_station_longitude  bike_id user_type  member_birth_year member_gender  
0            -122.415929       96  Customer             1987.0          Male  
1            -122.426614       88  Customer             1965.0        Female  
2            -122.267764     1094  Customer                NaN           NaN

In [3]:
# load in the dataset into a pandas dataframe, show first three rows
df_bike_1 = pd.read_csv('C:\\Users\\raz37388\\Desktop\\Udacity_assignment\\Assignment 4\\201801-fordgobike-tripdata.csv')
df_bike_2 = pd.read_csv('C:\\Users\\raz37388\\Desktop\\Udacity_assignment\\Assignment 4\\201802-fordgobike-tripdata.csv')
df_bike_3 = pd.read_csv('C:\\Users\\raz37388\\Desktop\\Udacity_assignment\\Assignment 4\\201803-fordgobike-tripdata.csv')
df_bike_4 = pd.read_csv('C:\\Users\\raz37388\\Desktop\\Udacity_assignment\\Assignment 4\\201804-fordgobike-tripdata.csv')
df_bike_5 = pd.read_csv('C:\\Users\\raz37388\\Desktop\\Udacity_assignment\\Assignment 4\\201805-fordgobike-tripdata.csv')

In [4]:
#combinning the datasets
df_combined = pd.concat([df_bike_1,df_bike_2, df_bike_3,df_bike_4, df_bike_5])
df_combined.head(2)

duration_sec                start_time                  end_time  \
0         75284  2018-01-31 22:52:35.2390  2018-02-01 19:47:19.8240   
1         85422  2018-01-31 16:13:34.3510  2018-02-01 15:57:17.3100   

   start_station_id                                 start_station_name  \
0               120                               Mission Dolores Park   
1                15  San Francisco Ferry Building (Harry Bridges Pl...   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.761420              -122.426435             285   
1               37.795392              -122.394203              15   

                                    end_station_name  end_station_latitude  \
0                         Webster St at O'Farrell St             37.783521   
1  San Francisco Ferry Building (Harry Bridges Pl...             37.795392   

   end_station_longitude  bike_id   user_type  member_birth_year  \
0            -122.431158     2765  Subscriber             1986.0   
1            -122.394203     2815    Customer                NaN   

  member_gender bike_share_for_all_trip  
0          Male                      No  
1           NaN                      No

In [5]:
#drop the bike share column 
df_combined.drop(['bike_share_for_all_trip'], axis=1, inplace= True)

In [6]:
#combine the data to obtain a dataset for the whole year for the SanFrancisso bay area
df = pd.concat([df_bike,df_combined])

In [7]:
# show first three rows
df.head(3)

duration_sec                start_time                  end_time  \
0         80110  2017-12-31 16:57:39.6540  2018-01-01 15:12:50.2450   
1         78800  2017-12-31 15:56:34.8420  2018-01-01 13:49:55.6170   
2         45768  2017-12-31 22:45:48.4110  2018-01-01 11:28:36.8830   

   start_station_id                                 start_station_name  \
0                74                              Laguna St at Hayes St   
1               284  Yerba Buena Center for the Arts (Howard St at ...   
2               245                             Downtown Berkeley BART   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.776435              -122.426244              43   
1               37.784872              -122.400876              96   
2               37.870348              -122.267764             245   

                                    end_station_name  end_station_latitude  \
0  San Francisco Public Library (Grove St at Hyde...             37.778768   
1                              Dolores St at 15th St             37.766210   
2                             Downtown Berkeley BART             37.870348   

   end_station_longitude  bike_id user_type  member_birth_year member_gender  
0            -122.415929       96  Customer             1987.0          Male  
1            -122.426614       88  Customer             1965.0        Female  
2            -122.267764     1094  Customer                NaN           NaN

In [8]:
df.shape

(1142896, 15)

In [9]:
#Find duplicates in the dataset
df[df.duplicated()]

Empty DataFrame
Columns: [duration_sec, start_time, end_time, start_station_id, start_station_name, start_station_latitude, start_station_longitude, end_station_id, end_station_name, end_station_latitude, end_station_longitude, bike_id, user_type, member_birth_year, member_gender]
Index: []

In [10]:
#Find duplicates -should return False since there is no duplicate
df.duplicated().any()

False

In [11]:
#Finding the null values
df.isnull().sum()

duration_sec                    0
start_time                      0
end_time                        0
start_station_id                0
start_station_name              0
start_station_latitude          0
start_station_longitude         0
end_station_id                  0
end_station_name                0
end_station_latitude            0
end_station_longitude           0
bike_id                         0
user_type                       0
member_birth_year          112634
member_gender              112370
dtype: int64

In [12]:
#drop the null values
df.dropna(inplace=True)

In [13]:
#Checking is there are any other Null value- should return 0 for all the feature
df.isnull().sum()

duration_sec               0
start_time                 0
end_time                   0
start_station_id           0
start_station_name         0
start_station_latitude     0
start_station_longitude    0
end_station_id             0
end_station_name           0
end_station_latitude       0
end_station_longitude      0
bike_id                    0
user_type                  0
member_birth_year          0
member_gender              0
dtype: int64

In [14]:
#Shape of the dataset
df.shape

(1030262, 15)

In [15]:
#Show first three rows
df.head(3)

duration_sec                start_time                  end_time  \
0         80110  2017-12-31 16:57:39.6540  2018-01-01 15:12:50.2450   
1         78800  2017-12-31 15:56:34.8420  2018-01-01 13:49:55.6170   
4         43603  2017-12-31 14:23:14.0010  2018-01-01 02:29:57.5710   

   start_station_id                                 start_station_name  \
0                74                              Laguna St at Hayes St   
1               284  Yerba Buena Center for the Arts (Howard St at ...   
4               239                      Bancroft Way at Telegraph Ave   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.776435              -122.426244              43   
1               37.784872              -122.400876              96   
4               37.868813              -122.258764             247   

                                    end_station_name  end_station_latitude  \
0  San Francisco Public Library (Grove St at Hyde...             37.778768   
1                              Dolores St at 15th St             37.766210   
4                          Fulton St at Bancroft Way             37.867789   

   end_station_longitude  bike_id   user_type  member_birth_year member_gender  
0            -122.415929       96    Customer             1987.0          Male  
1            -122.426614       88    Customer             1965.0        Female  
4            -122.265896     3167  Subscriber             1997.0        Female

## 2. Dataset Overview


### A. What is the structure of this dataset?

The data consisted of prices and attributes of approximately 1142896 passengers. After removing the null values the number of passenger is reduced to 1030262, which is large enough to make a sound analysis. There are 15 features in the dataset out of which 11 are numeric in natures. 

The qualitative variables provides information about the start station name, end_station_name, gender of the rider, and user type- whether the person is a Customer or a Subscriber. 


### B. What is/are the main feature(s) of interest in the dataset?

I'm most interested in figuring out how the bike riders make their journeys in diffrent parts of the day and whether it also depends on being subscriber or not. Also, I will investigate diffrent age group of the users and their riding patterns in terms of distance and duration covered during the journey. 

### C. What features in the dataset  will help support the investigation intothe feature(s) of interest?

To do the analysis, I will extract diffrents aspects of time such as- day, month, year, season and the age of the riders. For convenience of analysis I will also divide them into diffrernt age groups. Finally, I will analyze what aspects helps most in predicting how much distance a rider goes and what seems important for a person to be a subscriber or a customer. 

In [16]:
#data type information and non-null values
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1030262 entries, 0 to 179123
Data columns (total 15 columns):
duration_sec               1030262 non-null int64
start_time                 1030262 non-null object
end_time                   1030262 non-null object
start_station_id           1030262 non-null int64
start_station_name         1030262 non-null object
start_station_latitude     1030262 non-null float64
start_station_longitude    1030262 non-null float64
end_station_id             1030262 non-null int64
end_station_name           1030262 non-null object
end_station_latitude       1030262 non-null float64
end_station_longitude      1030262 non-null float64
bike_id                    1030262 non-null int64
user_type                  1030262 non-null object
member_birth_year          1030262 non-null float64
member_gender              1030262 non-null object
dtypes: float64(5), int64(4), object(6)
memory usage: 125.8+ MB


**Note: It can be obsrved that the start_time and end_time is in string format and for the sake of our analysis we need to convert it to date time to extract diffrent aspects of time from it.**

In [17]:
#converting the start and end time into date time format
df['start_time']= pd.to_datetime(df.start_time)
df['end_time']= pd.to_datetime(df.end_time)

In [18]:
# Creating the month category- will return the month of the year
df['month'] = df['start_time'].dt.strftime('%b')

# Extracting Day from the start-time
df['day'] = df['start_time'].dt.strftime('%a')

# finding the year from the start date
df['year'] = df['start_time'].dt.year

#finding starting month number- will return a value from 0-12
df['str_month_no'] = df['start_time'].dt.strftime('%m').astype(int)

#extracting hour- will return a value from 0 to 23 
df['start_hour'] = df['start_time'].dt.hour.astype(int)
df['end_hour'] = df['end_time'].dt.hour.astype(int)

#convert the duraiton of the journey in minute
df['duration_min'] = df['duration_sec']/60

In [19]:
#see first three rows to observe the changes
df.head(3)

duration_sec              start_time                end_time  \
0         80110 2017-12-31 16:57:39.654 2018-01-01 15:12:50.245   
1         78800 2017-12-31 15:56:34.842 2018-01-01 13:49:55.617   
4         43603 2017-12-31 14:23:14.001 2018-01-01 02:29:57.571   

   start_station_id                                 start_station_name  \
0                74                              Laguna St at Hayes St   
1               284  Yerba Buena Center for the Arts (Howard St at ...   
4               239                      Bancroft Way at Telegraph Ave   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.776435              -122.426244              43   
1               37.784872              -122.400876              96   
4               37.868813              -122.258764             247   

                                    end_station_name  end_station_latitude  \
0  San Francisco Public Library (Grove St at Hyde...             37.778768   
1                              Dolores St at 15th St             37.766210   
4                          Fulton St at Bancroft Way             37.867789   

       ...        user_type  member_birth_year member_gender  month  day  \
0      ...         Customer             1987.0          Male    Dec  Sun   
1      ...         Customer             1965.0        Female    Dec  Sun   
4      ...       Subscriber             1997.0        Female    Dec  Sun   

   year str_month_no  start_hour  end_hour  duration_min  
0  2017           12          16        15   1335.166667  
1  2017           12          15        13   1313.333333  
4  2017           12          14         2    726.716667  

[3 rows x 22 columns]

In [20]:
# A function to find whether the day is weekend or not
def find_weekend (df_column):
    
    """ This function takes the day column of a dataframe as an input and determine whether it is weeekend or not as an output.
    Saturday and Sunday will be marked as the Weekend while other days will be Weekdays"""
    
    if df_column == 'Sat':
        return 'Weekend'
    elif df_column == 'Sun':
        return 'Weekend'
    else:
        return 'Weekday'

In [21]:
# Applying the find_weekend function to the dataframe
df['day_type'] = df['day'].apply(find_weekend)

In [22]:
#Calculating age in age at the day of the journey
df['age'] = df['year']-df['member_birth_year'].astype(int)

In [23]:
#see first three rows to observe the changes of day_type and age
df.head(3)

duration_sec              start_time                end_time  \
0         80110 2017-12-31 16:57:39.654 2018-01-01 15:12:50.245   
1         78800 2017-12-31 15:56:34.842 2018-01-01 13:49:55.617   
4         43603 2017-12-31 14:23:14.001 2018-01-01 02:29:57.571   

   start_station_id                                 start_station_name  \
0                74                              Laguna St at Hayes St   
1               284  Yerba Buena Center for the Arts (Howard St at ...   
4               239                      Bancroft Way at Telegraph Ave   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.776435              -122.426244              43   
1               37.784872              -122.400876              96   
4               37.868813              -122.258764             247   

                                    end_station_name  end_station_latitude  \
0  San Francisco Public Library (Grove St at Hyde...             37.778768   
1                              Dolores St at 15th St             37.766210   
4                          Fulton St at Bancroft Way             37.867789   

  ...   member_gender  month  day  year str_month_no start_hour end_hour  \
0 ...            Male    Dec  Sun  2017           12         16       15   
1 ...          Female    Dec  Sun  2017           12         15       13   
4 ...          Female    Dec  Sun  2017           12         14        2   

   duration_min  day_type  age  
0   1335.166667   Weekend   30  
1   1313.333333   Weekend   52  
4    726.716667   Weekend   20  

[3 rows x 24 columns]

From the source [1] it can be seen the the San Francisco bay area has four seasons in the year. Hence, the Seasons information will now be extracted from the dataset. 

**Season-1.Summer: June to August**

**Season-2.Autumn: September to November**

**Season-3.Winter: December to February**

**Season-4.Spring: March to May**


** [1].https://www.studentflights.com.au/destinations/san-francisco/weather**

In [35]:
def find_season (df_column):
    
    """ This function takes the day column of a dataframe as an input and determine whether it is weeekend or not as an output.
    Saturday and Sunday will be marked as the Weekend while other days will be Weekdays"""
    
    if df_column >5 and df_column <9:
        return 'Summer'
    
    elif df_column >8 and df_column <12:
        return 'Autumn'
    
    elif df_column >2 and df_column <6:
        return 'Spring'

    else:
        return 'Winter'

In [25]:
# Applying the find_weekend function to the dataframe
df['season']= df['str_month_no'].apply(find_season)

# Find the number of unique Seasons- should return the four season
df['season'].unique()

array(['Winter', 'Autumn', 'Summer', 'Spring'], dtype=object)

In [26]:
#Show the dataset to observe the Change
df.head(1)

duration_sec              start_time                end_time  \
0         80110 2017-12-31 16:57:39.654 2018-01-01 15:12:50.245   

   start_station_id     start_station_name  start_station_latitude  \
0                74  Laguna St at Hayes St               37.776435   

   start_station_longitude  end_station_id  \
0              -122.426244              43   

                                    end_station_name  end_station_latitude  \
0  San Francisco Public Library (Grove St at Hyde...             37.778768   

    ...    month  day  year  str_month_no start_hour end_hour duration_min  \
0   ...      Dec  Sun  2017            12         16       15  1335.166667   

   day_type  age  season  
0   Weekend   30  Winter  

[1 rows x 25 columns]

**Note: We need to calcuate the distance travelled from the informaiton of Longitude and Latitude. For that er need to import a new library.The code to calculate the distance is take from Ref-2**

[2]. Calcute Distance Between two points: https://stackoverflow.com/questions/44446862/calculate-distance-between-latitude-and-longitude-in-dataframe

In [27]:
from geopy.distance import vincenty
def distance_calc (row):
    "this function calculate the distance between the starting and the ending station from the information of longitude and laitude and returns the distance in meters."
    start = (row['start_station_latitude'], row['start_station_longitude'])
    stop = (row['end_station_latitude'], row['end_station_longitude'])

    return vincenty(start, stop).meters/100

In [28]:
df['distance'] = df.apply (lambda row: distance_calc (row),axis=1)

In [29]:
#Check if there is any 0 distance. 
df.query('distance==0')

duration_sec              start_time                end_time  \
6               4507 2017-12-31 23:49:28.422 2018-01-01 01:04:35.619   
64               547 2017-12-31 22:52:35.849 2017-12-31 23:01:43.178   
76              2646 2017-12-31 22:02:13.306 2017-12-31 22:46:19.330   
86              1961 2017-12-31 21:59:37.081 2017-12-31 22:32:18.498   
187             1609 2017-12-31 19:50:52.507 2017-12-31 20:17:41.997   
219              165 2017-12-31 19:50:30.763 2017-12-31 19:53:16.559   
310             1596 2017-12-31 17:35:13.180 2017-12-31 18:01:49.197   
326              223 2017-12-31 17:45:56.753 2017-12-31 17:49:40.319   
349            10509 2017-12-31 14:38:46.875 2017-12-31 17:33:56.394   
357             2158 2017-12-31 16:55:21.414 2017-12-31 17:31:19.507   
371             4857 2017-12-31 16:02:27.299 2017-12-31 17:23:24.673   
442             2391 2017-12-31 16:04:11.514 2017-12-31 16:44:03.289   
467              972 2017-12-31 16:17:05.314 2017-12-31 16:33:18.035   
469              868 2017-12-31 16:18:10.076 2017-12-31 16:32:38.720   
490             2191 2017-12-31 15:47:41.948 2017-12-31 16:24:13.466   
492             2179 2017-12-31 15:47:42.334 2017-12-31 16:24:01.982   
494             2613 2017-12-31 15:36:15.249 2017-12-31 16:19:49.124   
495             2615 2017-12-31 15:36:12.470 2017-12-31 16:19:47.941   
517               81 2017-12-31 16:09:26.598 2017-12-31 16:10:48.229   
525              296 2017-12-31 15:58:47.884 2017-12-31 16:03:44.518   
546             1219 2017-12-31 15:38:23.721 2017-12-31 15:58:42.751   
558             5012 2017-12-31 14:27:47.251 2017-12-31 15:51:19.662   
559             4649 2017-12-31 14:33:44.758 2017-12-31 15:51:13.868   
585              131 2017-12-31 15:40:30.671 2017-12-31 15:42:41.955   
660             7439 2017-12-31 13:05:24.278 2017-12-31 15:09:23.845   
662              235 2017-12-31 15:03:48.971 2017-12-31 15:07:44.415   
697              256 2017-12-31 14:49:19.868 2017-12-31 14:53:36.413   
782             1307 2017-12-31 13:59:12.382 2017-12-31 14:20:59.683   
784             1199 2017-12-31 13:59:05.691 2017-12-31 14:19:04.869   
827              768 2017-12-31 13:53:51.051 2017-12-31 14:06:39.955   
...              ...                     ...                     ...   
177057           159 2018-05-01 10:01:46.490 2018-05-01 10:04:26.157   
177061           184 2018-05-01 10:01:09.103 2018-05-01 10:04:13.317   
177063           198 2018-05-01 10:00:42.723 2018-05-01 10:04:01.101   
177068           183 2018-05-01 09:59:58.324 2018-05-01 10:03:02.292   
177072           127 2018-05-01 10:00:42.523 2018-05-01 10:02:50.265   
177075            97 2018-05-01 10:00:43.475 2018-05-01 10:02:20.627   
177090           134 2018-05-01 09:57:34.543 2018-05-01 09:59:48.860   
177091           120 2018-05-01 09:57:34.335 2018-05-01 09:59:34.428   
177121           244 2018-05-01 09:52:24.227 2018-05-01 09:56:28.530   
177155          3026 2018-05-01 09:02:16.229 2018-05-01 09:52:42.582   
177174           244 2018-05-01 09:45:52.825 2018-05-01 09:49:57.589   
177317            79 2018-05-01 09:32:58.932 2018-05-01 09:34:18.172   
177417          1551 2018-05-01 08:59:58.845 2018-05-01 09:25:50.150   
178052           129 2018-05-01 08:39:27.914 2018-05-01 08:41:37.093   
178126          1009 2018-05-01 08:19:25.636 2018-05-01 08:36:15.278   
178572           123 2018-05-01 07:54:53.058 2018-05-01 07:56:56.937   
178575           126 2018-05-01 07:54:41.243 2018-05-01 07:56:47.416   
178583           101 2018-05-01 07:54:21.937 2018-05-01 07:56:03.760   
178586           112 2018-05-01 07:54:01.570 2018-05-01 07:55:53.724   
178621            83 2018-05-01 07:52:04.091 2018-05-01 07:53:27.978   
178622           164 2018-05-01 07:50:41.396 2018-05-01 07:53:25.714   
178634            98 2018-05-01 07:50:59.068 2018-05-01 07:52:37.787   
178648           465 2018-05-01 07:44:01.269 2018-05-01 07:51:46.617   
178666           149 2018-05-01 07:46:52.264 201

In [30]:
# Drop rows with 0 Values
df = df.loc[df.ne(0).all(axis=1)]

In [31]:
#Check if there is any 0 distance. 
df.query('distance==0')

Empty DataFrame
Columns: [duration_sec, start_time, end_time, start_station_id, start_station_name, start_station_latitude, start_station_longitude, end_station_id, end_station_name, end_station_latitude, end_station_longitude, bike_id, user_type, member_birth_year, member_gender, month, day, year, str_month_no, start_hour, end_hour, duration_min, day_type, age, season, distance]
Index: []

[0 rows x 26 columns]

In [32]:
df.head()

duration_sec              start_time                end_time  \
0          80110 2017-12-31 16:57:39.654 2018-01-01 15:12:50.245   
1          78800 2017-12-31 15:56:34.842 2018-01-01 13:49:55.617   
4          43603 2017-12-31 14:23:14.001 2018-01-01 02:29:57.571   
28           871 2017-12-31 23:37:47.192 2017-12-31 23:52:18.473   
32           733 2017-12-31 23:38:38.861 2017-12-31 23:50:52.504   

    start_station_id                                 start_station_name  \
0                 74                              Laguna St at Hayes St   
1                284  Yerba Buena Center for the Arts (Howard St at ...   
4                239                      Bancroft Way at Telegraph Ave   
28               323                                 Broadway at Kearny   
32               323                                 Broadway at Kearny   

    start_station_latitude  start_station_longitude  end_station_id  \
0                37.776435              -122.426244              43   
1                37.784872              -122.400876              96   
4                37.868813              -122.258764             247   
28               37.798014              -122.405950              14   
32               37.798014              -122.405950              14   

                                     end_station_name  end_station_latitude  \
0   San Francisco Public Library (Grove St at Hyde...             37.778768   
1                               Dolores St at 15th St             37.766210   
4                           Fulton St at Bancroft Way             37.867789   
28                              Clay St at Battery St             37.795001   
32                              Clay St at Battery St             37.795001   

      ...      day  year str_month_no  start_hour end_hour duration_min  \
0     ...      Sun  2017           12          16       15  1335.166667   
1     ...      Sun  2017           12          15       13  1313.333333   
4     ...      Sun  2017           12          14        2   726.716667   
28    ...      Sun  2017           12          23       23    14.516667   
32    ...      Sun  2017           12          23       23    12.216667   

   day_type  age  season   distance  
0   Weekend   30  Winter   9.448761  
1   Weekend   52  Winter  30.711442  
4   Weekend   20  Winter   6.377721  
28  Weekend   38  Winter   6.238980  
32  Weekend   37  Winter   6.238980  

[5 rows x 26 columns]

In [34]:
df.shape

(1001173, 26)

In [33]:
# Store the dataset
%store df

Stored 'df' (DataFrame)


**Note: Now the dataset is ready to perform all the analysis as we have done some primary wrangling. However, we will continue to wrangle data based on need.**